Support Hawei Cloud (PFS && OBS)

yabola opened the pull request:
1. Support Huawei Cloud PFS( parallel file system )
2. Move Huawei Cloud OBS from Alluxio-extension. https://github.com/Alluxio/alluxio-extensions/pull/32
3. fix some UT && add doc for OBS/PFS

PFS is an extension of OBS. 

1. The main difference is  dir in PFS will be explicitly created . But other storage (like s3,obs....) won't create a real dir.
(when PFS,  `OBSUnderFileSystem#getObjectStatus`will return meta in dir , but we want null.).
This will cause dir be considered as a file in some conditions and files in this dir won't be listed.

2.  Assuming this scenario: 
    `/xx/file1/` ( File1 actually exists, which is a file) , but we should pay attention, there is `/` after file1 name.
     When OBS , the path object meta is null.
     When PFS, the path object meta is not null. The object meta is same as `/xx/file1`
     
3.  PFS supports `Rename` directory , OBS doesn't support.

4. Assuming this scenario: 
   There are several files as follows: /dir1, /dir1/file1
   When OBS (also s3...) ,   delete `/dir1` ---> dir1 & file1 all exists.   Then delete `/dir1/file1`, dir1 & file1 all are deleted. 
   When PFS ,  delete `/dir1` ---> dir1 & file1 all exists.   Then delete `/dir1/file1`, file1 is deleted, dir1 is not deleted.
   This means that when PFS, We should first  delete `/dir1/file1`, then delete `/dir1`.

LuQQiu commented on underfs/obs/src/main/resources/META-INF/services/alluxio.underfs.UnderFileSystemFactory:
```diff
@@ -0,0 +1,12 @@
+# See the NOTICE file distributed with this work for information regarding copyright ownership.
+#
+
+alluxio.underfs.obs.OBSUnderFileSystemFactory
```
add an extra line at the end of file

LuQQiu reviewed (commented):
@yabola can you help modify the OBS doc and possible add PFS doc?
Can you help fix the test failures?

Appreciate your contribution to PFS/OBS!

yabola commented:
retest this please

yabola commented:
> @yabola can you help modify the OBS doc and possible add PFS doc?
> Can you help fix the test failures?
> 
> Appreciate your contribution to PFS/OBS!

It seems Jenkins can’t pull https://repo.huaweicloud.com/repository/maven/huaweicloudsdk . How can I retest this in Jenkins?

yabola commented:
alluxio-bot, test this please

yabola commented:
@LuQQiu 
I have added the UT and modified related documents.
But It seems Jenkins can’t pull https://repo.huaweicloud.com/repository/maven/huaweicloudsdk

LuQQiu commented:
@yabola Yeah it cannot pull Huawei dependencies. if it cannot pull, we cannot put it in the main repo since it will prevent us from building the release tarballs. If there are no other ways of getting the dependencies, we can only leave it in the alluxio-extension github repo and hope that users are able to get the dependency when they trying to build the jar

yabola commented:
> @LuQQiu I found another maven dependency SDK for OBS client (provided by huawei ). It had passed in Jenkins.
> 
> `Failed to execute goal com.mycila:license-maven-plugin:4.0.rc1:check (default) on project alluxio-underfs-obs: Some files do not have the expected license header -> [Help 1]`
> I don't know the reason for this.

Warning: 472 [WARNING] Missing header in: /usr/src/alluxio/underfs/obs/pom.xml
@LuQQiu  I had fixed it.

LuQQiu commented:
@yabola That's amazing! Thanks for fixing it!

LuQQiu commented on docs/en/ufs/OBS.md:
```diff
@@ -76,6 +55,13 @@ Valid values are similar to `obs.cn-east-2.myhuaweicloud.com` and `obs.cn-south-
 Available endpoints are listed in the [Region and Endpoint list](https://developer.huaweicloud.com/en-us/endpoint)
 under the Object Storage Service category.
 
+If you want to use the `PFS` parallel file system, you need to configure in `conf/alluxio-site.properties` as below. The rest of the operations are the same as the `OBS` file system. Add:
```
you need to configure in `conf/alluxio-site.properties` as below.
-> please add the following configuration in `conf/alluxio-site.properties`.

LuQQiu commented on underfs/obs/src/test/java/alluxio/underfs/obs/OBSOutputStreamTest.java:
```diff
@@ -0,0 +1,195 @@
+  @PrepareForTest(OBSOutputStream.class)
+  public void testWrite2() throws Exception {
+    PowerMockito.whenNew(BufferedOutputStream.class)
+            .withArguments(Mockito.any(DigestOutputStream.class)).thenReturn(mLocalOutputStream);
```
please fix the indentations, thanks

LuQQiu commented on underfs/obs/pom.xml:
```diff
@@ -0,0 +1,75 @@
+            </plugin>
+        </plugins>
+    </build>
+</project>
```
add a new line at the end of file

LuQQiu commented on underfs/obs/src/main/java/alluxio/underfs/obs/OBSPropertyKey.java:
```diff
@@ -0,0 +1,43 @@
+ * OBS configuration property keys.
+ */
+@ThreadSafe
+public final class OBSPropertyKey {
```
Please move all the property key into PropertyKey file and delete this file.
This file is used in ufs-extensions but not needed when move back to Alluxio

yabola replied:
done

LuQQiu reviewed (commented):
Thanks for moving it back to main repo! left some comments
Please redo the runUfsTests for both OBS and PFS, thanks!

LuQQiu commented:
@apc999 FYI

yabola commented:
@LuQQiu  I  had retested `./bin/alluxio runTests` and `runUfsTests` script in both OBS and PFS. Only failed at `getModTimeTest` ( I am using Huawei Cloud in Thailand).

LuQQiu commented on underfs/obs/src/main/java/alluxio/underfs/obs/OBSInputStream.java:
```diff
@@ -58,29 +66,29 @@
-   * @param client the client for OSS
-   * @param retryPolicy retry policy in case the key does not exist
+   * @param bucketName          the name of the bucket
+   * @param key                 the key of the file
```
revert the changes?

LuQQiu commented on underfs/obs/src/main/java/alluxio/underfs/obs/OBSInputStream.java:
```diff
@@ -58,29 +66,29 @@
    */
   OBSInputStream(String bucketName, String key, ObsClient client, RetryPolicy retryPolicy,
-      long multiRangeChunkSize) throws IOException {
+                 long multiRangeChunkSize) throws IOException {
```
revert the changes?

LuQQiu commented on underfs/obs/src/main/java/alluxio/underfs/obs/OBSInputStream.java:
```diff
@@ -58,29 +66,29 @@
-   * @param retryPolicy retry policy in case the key does not exist
+   * @param bucketName          the name of the bucket
+   * @param key                 the key of the file
+   * @param client              the OBS client
```
revert?

LuQQiu commented on underfs/obs/src/main/java/alluxio/underfs/obs/OBSInputStream.java:
```diff
@@ -58,29 +66,29 @@
    */
   OBSInputStream(String bucketName, String key, ObsClient client, long position,
-      RetryPolicy retryPolicy, long multiRangeChunkSize) throws IOException {
+                 RetryPolicy retryPolicy, long multiRangeChunkSize) throws IOException {
```
revert?

LuQQiu commented on underfs/obs/src/main/java/alluxio/underfs/obs/OBSOutputStream.java:
```diff
@@ -121,7 +135,7 @@ public void write(byte[] b) throws IOException {
    * output stream. Before close, the bytes are all written to local file.
    *
-   * @param b the byte array
+   * @param b   the byte array
```
revert?

LuQQiu commented on underfs/obs/src/main/java/alluxio/underfs/obs/OBSUnderFileSystem.java:
```diff
@@ -49,36 +49,44 @@
-    String secretKey = conf.get(OBSPropertyKey.OBS_SECRET_KEY);
-    String endPoint = conf.get(OBSPropertyKey.OBS_ENDPOINT);
-    String bucketType = conf.get(OBSPropertyKey.OBS_BUCKET_TYPE);
+                                                  UnderFileSystemConfiguration conf) throws Exception {
```
Indentation?

LuQQiu commented on underfs/obs/src/main/java/alluxio/underfs/obs/OBSOutputStream.java:
```diff
@@ -46,33 +46,47 @@
    */
   public OBSOutputStream(String bucketName, String key, ObsClient client,
-      List<String> tmpDirs) throws IOException {
+                         List<String> tmpDirs) throws IOException {
```
Revert?

LuQQiu reviewed (approved):
Left some comments about code style
our indentation is 2 spaces, for separating line it's 4 spaces
Please fix the codestyle issue, otherwise LGTM
thanks for testing out

yabola commented:
@LuQQiu Thanks for the detailed review. I change my IDE configuration ( 
`Method declaration parameters` ->  `align when multiline`) and reformat my code.

LuQQiu commented on underfs/obs/src/main/java/alluxio/underfs/obs/OBSConstants.java:
```diff
@@ -0,0 +1,22 @@
+ */
+@ThreadSafe
+public final class OBSConstants {
+  public static final String HEADER_OBS = "obs://";
```
Sorry one more thing!
Can you remove this file and put this header into `alluxio.Constants` together with other ufs like s3://

yabola replied:
done~

LuQQiu reviewed (commented):
Left one minor comments, otherwise this PR is ready to be merged

LuQQiu commented:
alluxio-bot, merge this please

LuQQiu commented:
@yabola can you open a new PR to give more information about what is PFS VS OBS in alluxio documentation?

yabola commented:
> @yabola can you open a new PR to give more information about what is PFS VS OBS in alluxio documentation?

@LuQQiu Ok~ What points need to be said? like PR description?
